api/wft-complete-metadata.md

```diff
@@ -0,0 +1,126 @@
+state.
+
+## Use cases
+
+### Reporting currently-being-awaited commands & facts about them
+
+Right now, besides a query, there's no simple way for a user to understand what their workflow is
+actively waiting on. Solving this would make it much easier for users to have insight into why
+their workflow is or isn't making progress, and could be nicely exposed in the UI.
+
+In particular, understanding what's going on while your workflow is currently running local
+activities and performing WFT heartbeats is very opaque. This would make that much more clear.
+
+Imagine being able to get information like this (and do useful things with it in the UI):
+
+```
+waiting_on: [
+    { command_type: "local_activity", activity_name: "la_func", id: 1, retries: 0, location: "my_file.py:23" },
+    { command_type: "local_activity", activity_name: "la_func", id: 2, retries: 5, location: "my_file.py:23" },
+    { command_type: "timer", id: 3, duration: "5s", location: "my_file.py:42" },
+]
+```
+
+### SDK internal versioning
+
+We would like to be able to make SDK changes which affect how history is written, which means
+flagging our behavior based on what version of the SDK the history was produced with. Note: *not*
+relating to concepts of versioning a *user's* code, but rather the SDK itself.
+
+This internal version could easily slot into this metadata.
+
+Note: This could potentially go in `temporal.api.taskqueue.v1.VersionId` too, but that's more
+intended to relate to user code.
+
+### Dynamic config adjustments
+
+Allow the worker to set configurable things which don't directly impact determinism or workflow
+state. For example, changing the WFT timeout for the next task.
+
+### Various debugging / analysis information
+
+There's a whole world of little bits of info that could be useful to enable selectively to help
+users debug their workflows, help us learn about performance information, etc.
+
+Some examples:
+* Emit the OTel span id that was associated with the WFT - now the user can go easily look up any
+  logs/events that were emitted as part of that WFT.
+* Metrics work great for all sorts of latency measurements, but what if you wanted to know about
+  how long a _specific_ WFT took? You could emit that.
+* Interceptors could pipe whatever arbitrary data the user wants into there, for use cases where
+  memo / search attributes / side effects may not be appropriate.
+
+### Things we haven't thought of
+
+Having this metadata be suitably generic means we dont need to update the API repo for every
+possible usecase. Particularly ones where the SDK is the only one who needs to be able to read it.
+
+## Implementation Proposal
+
+We'd add a new metadata field to the
+[WFT response](https://github.com/temporalio/api/blob/master/temporal/api/workflowservice/v1/request_response.proto#L274).
+And the [WFT completed event](https://github.com/temporalio/api/blob/f0350f8032ad2f0c60c539b3b61ea37f412f1cf7/temporal/api/history/v1/message.proto#L186).
```

Comment:
Adding arbitrarily-sized attributes into history exposes us to resource exhaustion. For example, we had a major outage caused by storing too many child workflow executions in the mutable state for executions.

Reply:
I resolved the other comments since they all largely resolve around the same concern w/ tracing.

Yep, bloat is called out in possible drawbacks. 

Not everyone uses tracing (in fact, I'd guess most users don't). This potentially gives them a way to write data into a place that we currently provide visibility most easily to them. Not everything makes sense to write as tracing data either.

Beyond that, this proposal isn't saying we'll definitely do that. It's something it could support if we ever needed to, now. No user data is going to be written here until if and when we decide we want to do that. If we did, almost certainly we'd put limits on that like we do w/ existing user-created fields.